### ISSUE TITLE
Question about `q_pred_one_timestep`

### ISSUE THREAD

**Issue by cantabile-kwok** [ehoogeboom/multinomial_diffusion#9]
Thanks for the brilliant work! I really have some trouble understanding the function `q_pred_one_timestep`, although it has been illustrated in the published paper.

> Some magic is happening in q_pred_one_timestep. Recall that at some point we need to compute $\mathcal C(x_t|(1 − \beta_t)x_{t-1}+ \beta_t/K)$ for different values of $x_t$, which when treated as a function outputs $(1 − \beta_t) + \beta_t/K$ if $x_t=x_{t-1}$ and $\beta_t/K$ otherwise. This function is symmetric, meaning that $\mathcal C(x_t|(1 − \beta_t)x_{t-1} + \beta_t/K) = \mathcal C(x_{t-1}|(1 − \beta_t)x_t + \beta_t/K)$. This is why we can switch the conditioning and immediately return the different probability vectors for $x_t$. This also corresponds to Equation 13.

My question is, although the equation $\mathcal C(x_t|(1 − \beta_t)x_{t-1} + \beta_t/K) = \mathcal C(x_{t-1}|(1 − \beta_t)x_t + \beta_t/K)$ holds well, this is for explicit $x_t$ and $x_{t-1}$. In other words, only if we has already determined the value of $x_t$ and $x_{t-1}$, can we write down such an equation. In the function `q_pred_one_timestep`, you provide $x_t$ into the function, and compute $\mathcal C((1-\beta_t)x_t + \beta_t/K)$ as a whole distribution, which I think does not equal to the distribution $\mathcal C((1-\beta_t)x_{t-1} + \beta_t/K)$.
In other words, what confuses me is that, if we want to get $q(x_t|x_{t-1})$, isn't it necessary to determine $x_{t-1}$ first? In this function, we have provided $x_t$, what is the point of calculating its distribution then?

Any reminder is greatly appreciated!

**Comment by cantabile-kwok** [comment]
Hmmm, an interesting thing to notice is, although the function seems skeptical with respect to its name, it does not hurt the other parts of the code. This is because this function is only called in `q_posterior`, where the purpose of `q_pred_one_timestep` is actually to give $(1-\beta_t)x_t + \beta_t/K$, rather than $q(x_{t}|x_{t-1})=\mathcal C(x_t | (1-\beta_t)x_{t-1} +\beta_t/K)$. So this function still performs the right job, although not actually "pred one timestep".

**Comment by cantabile-kwok** [comment]
Before this issue is closed, I want to ask another question:
In the paper, as in Eq.(14), you parametrize the model with $\hat x_0 = \mu(x_t,t)$. In other words, the model predicts a deterministic $x_0$. But in the code, the model actually predicts a distribution of $x_0$, i.e. $p_\theta(\hat x_0\mid x_t)$. Also, I infer from some of the variables' names (`log_EV_qxtmin_x0` etc.) that you are dealing with expectation probably with respect to $\hat x_0$, so that the model truly predicts a distribution of $\hat x_0$.

If this is the actual case, then the KL term should be re-considered, because $p_\theta(x_{t-1}\mid x_t)=\Sigma_{\hat x_0=1}^K p_\theta(\hat x_0\mid x_t)q(x_{t-1}\mid x_t, \hat x_0)$. The $q(x_{t-1}\mid x_t, \hat x_0)$ could be processed by Eq.(13). However, the summation is outside everything. Meanwhile, in the code `q_posterior`, the normalization is done at the last step. I have some trouble understanding this, as the normalizing factor $\Sigma_{k=1}^K \tilde \theta_k$ is relevant to $\hat x_0$ also. You certainly cannot take summation of the numerator and denominator respectively. 

I guess there is also some magic behind them, but I really couldn't figure out. I noticed there is [another issue](https://github.com/ehoogeboom/multinomial_diffusion/issues/7) talking about this, but I didn't find solution to my question.

**Comment by ehoogeboom** [comment]
Hi, 

Regarding the first question, the function q_pred_one_timestep returns a probability vector. This keeps working if the conditioning is a probability vector.

Regardless of distributions, it’s probably easiest to verify that the function returns the right values. Somewhat confusingly, x_hat_0 is a probability vector. Not a single category. I hope we mentioned that in the paper though.

The other GitHub issue you refer to talks about an edge case of the first time step. I don’t think it’s directly related to this.

**Comment by cantabile-kwok** [comment]
Hi Emiel! 

Thank you for taking look at these questions. Regarding the first question, I still have some confusion. I agree that "it keeps working if the conditioning is a probability vector". My worry is that although $\mathcal C(x_t|(1 − \beta_t)x_{t-1} + \beta_t/K) = \mathcal C(x_{t-1}|(1 − \beta_t)x_t + \beta_t/K)$ holds, it does not imply $(1 − \beta_t)x_{t-1} + \beta_t/K = (1 − \beta_t)x_t + \beta_t/K$. If the function `q_pred_one_timestep` is designed to return $q_{t\mid t-1}(\cdot\mid x_{t-1})$ (which is a probability vector), then it should strictly return $(1 − \beta_t)x_{t-1} + \beta_t/K$, regardless of whether $x_t$ is one-hot or a probability vector, right?

**Comment by ehoogeboom** [comment]
Hi, 

Happy to help :). I think that when you work out the product of the two distributions for every possible x_t-1, meaning:

q(x_t | x_t-1) q(x_t-1 | x_0) as a vector indexed by x_t-1, then one finds that the first term simplifies to [ (1 - beta_t) 1[x_t-1 = x_t] + beta_t / K ] and the second term simplifies to [alpha_bar_t-1 1[x_0 = x_t-1] + (1 - alpha_bar_t-1) / K], where 1[...] are indicator functions. Let me know if that works!

**Comment by cantabile-kwok** [comment]
Hmm yes, I guess that is the same with the illustrations given in the paper. Anyway the first question does not influence the correctness of the code : ) I will focus on figuring out the second one, appreciate the help!

**Comment by ehoogeboom** [comment]
It may also be worth having a look at https://arxiv.org/abs/2107.03006, who expanded on this

**Comment by cantabile-kwok** [comment]
I think I can have a better description about my second question now.

First of all, the function [`q_pred`](https://github.com/ehoogeboom/multinomial_diffusion/blob/9d907a60536ad793efd6d2a6067b3c3d6ba9fce7/diffusion_utils/diffusion_multinomial.py#L149)  actually outputs **an expectation** $\mathbb E[q(x_{t}\mid x_0)]$, where the expectation is taken with respect to the distribution of $x_0$, specified by argument `log_x_start`. When `log_x_start` is a log-one-hot vector, it reduces to $q(x_t\mid x_0)$ as usual, where $x_0$ represents a single category instead of a distribution vector.

Second, [`q_pred` is called in `q_posterior` with argument `t-1`](https://github.com/ehoogeboom/multinomial_diffusion/blob/9d907a60536ad793efd6d2a6067b3c3d6ba9fce7/diffusion_utils/diffusion_multinomial.py#L178), so that we know variable `log_EV_qxtmin_x0` stands for $\mathbb E[q(x_{t-1}\mid x_0)]$, the same with its name. Then [L187](https://github.com/ehoogeboom/multinomial_diffusion/blob/9d907a60536ad793efd6d2a6067b3c3d6ba9fce7/diffusion_utils/diffusion_multinomial.py#L187) adds this with `self.q_pred_one_timestep(log_x_t, t)`, before normalization. This process can be summarized by this equation (in https://beckham.nz/2022/07/11/d3pms.html):
<img width="785" alt="image" src="https://user-images.githubusercontent.com/58417810/204747054-009188a5-ec93-495f-880b-32fea91e1ee0.png">
Meanwhile, the transition from Eq.(9) to Eq.(10) above is what I feel doubtful. The proportionality may not hold exactly. This is because
<img width="675" alt="image" src="https://user-images.githubusercontent.com/58417810/204746378-133df5ee-bebe-486a-bea0-d6869a6118e0.png">

By performing some small coding tests, I verified that `q_posterior` does not perfectly match the true posterior distribution, but only to a small difference. I think it also works because the difference is somewhat small in scale?

**Comment by LZhengisme** [comment]
> I think I can have a better description about my second question now.
> 
> First of all, the function [`q_pred`](https://github.com/ehoogeboom/multinomial_diffusion/blob/9d907a60536ad793efd6d2a6067b3c3d6ba9fce7/diffusion_utils/diffusion_multinomial.py#L149) actually outputs **an expectation** E[q(xt∣x0)], where the expectation is taken with respect to the distribution of x0, specified by argument `log_x_start`. When `log_x_start` is a log-one-hot vector, it reduces to q(xt∣x0) as usual, where x0 represents a single category instead of a distribution vector.
> 
> Second, [`q_pred` is called in `q_posterior` with argument `t-1`](https://github.com/ehoogeboom/multinomial_diffusion/blob/9d907a60536ad793efd6d2a6067b3c3d6ba9fce7/diffusion_utils/diffusion_multinomial.py#L178), so that we know variable `log_EV_qxtmin_x0` stands for E[q(xt−1∣x0)], the same with its name. Then [L187](https://github.com/ehoogeboom/multinomial_diffusion/blob/9d907a60536ad793efd6d2a6067b3c3d6ba9fce7/diffusion_utils/diffusion_multinomial.py#L187) adds this with `self.q_pred_one_timestep(log_x_t, t)`, before normalization. This process can be summarized by this equation (in https://beckham.nz/2022/07/11/d3pms.html): <img alt="image" width="785" src="https://user-images.githubusercontent.com/58417810/204747054-009188a5-ec93-495f-880b-32fea91e1ee0.png"> Meanwhile, the transition from Eq.(9) to Eq.(10) above is what I feel doubtful. The proportionality may not hold exactly. This is because <img alt="image" width="675" src="https://user-images.githubusercontent.com/58417810/204746378-133df5ee-bebe-486a-bea0-d6869a6118e0.png">
> 
> By performing some small coding tests, I verified that `q_posterior` does not perfectly match the true posterior distribution, but only to a small difference. I think it also works because the difference is somewhat small in scale?

Hi, 

I was having the same problem before. I think you are right that this equality does not hold in general since $q(x_{t-1} | x_t, x_0)$ is NOT linear w.r.t. $x_0$. 

However, the authors of this work did not claim $p_\theta(x_{t-1}\mid x_t)=\Sigma_{\hat x_0=1}^K p_\theta(\hat x_0\mid x_t)q(x_{t-1}\mid x_t, \hat x_0)$ in the paper, although some following-ups assume this holds (but note that it does hold in the case of absorbing diffusion, as proposed in [D3PMs](https://arxiv.org/abs/2107.03006)). Multinomial diffusion here just substitutes $x_0$ in $q(x_{t-1} | x_t, x_0)$ with a soft prediction $\tilde{x}\_0$ to define our generative process $p_{\theta}(x_{t-1}|x_t)$, and that's all. I hope this could make things a little bit clearer.

**Comment by cantabile-kwok** [comment]
@LZhengisme Thanks for pointing out something that I also automatically assumed to be correct! Though I am a little bit confused about why it may not hold, and why "substitutes $x_0$ in $q(x_{t-1}|x_t, x_0)$ with a soft prediction $\tilde x_0$" is different from that assumption. Could you please kindly elaborate on that a bit more?

**Comment by LZhengisme** [comment]
@cantabile-kwok  In my opinion, either "direct substitution" or $p_\theta(x_{t-1}\mid x_t)=\Sigma_{\hat x_0=1}^K p_\theta(\hat x_0\mid x_t)q(x_{t-1}\mid x_t, \hat x_0)$ **are just two different ways to parameterize the generative model $p_\theta(x_{t-1}\mid x_t)$**. We can define our generative model in various ways, which should be inherently different and enjoy different properties (although they might also happen to be equivalent in some cases). 

To see the difference between these two approaches in formulation, let's assume the expectation $\mathbb{E}$ is taken over model prediction $p\_{\theta}(\hat{x}\_0\mid x_t)$. Denote $\hat{x}\_0$ as **one-hot** variables and $\widetilde{x}\_0 = E[\hat{x}\_0]$ as the **soft** probability vector.
- "Direct substitution" can be understood as $q(x_{t-1}\mid x_t, \widetilde{x}\_0) = q(x_{t-1}\mid x_t, \mathbb E[\hat{x}\_0])$;
- The other way you referred to is actually $p_\theta(x_{t-1}\mid x_t)=\Sigma_{\hat x_0=1}^K p_\theta(\hat x_0\mid x_t)q(x_{t-1}\mid x_t, \hat x_0) = \mathbb{E}[q(x_{t-1}\mid x_t, \hat{x}\_0)]$.

These two terms are not equal in general, and the reason is precisely the same as what you have pointed out in your [comment](https://github.com/ehoogeboom/multinomial_diffusion/issues/9#issuecomment-1331795544):

> Meanwhile, the transition from Eq.(9) to Eq.(10) above is what I feel doubtful. The proportionality may not hold exactly. This is because ...

**Comment by cantabile-kwok** [comment]
> @cantabile-kwok In my opinion, either "direct substitution" or pθ(xt−1∣xt)=Σx^0=1Kpθ(x^0∣xt)q(xt−1∣xt,x^0) **are just two different ways to parameterize the generative model pθ(xt−1∣xt)**. We can define our generative model in various ways, which should be inherently different and enjoy different properties (although they might also happen to be equivalent in some cases).
> 
> To see the difference between these two approaches in formulation, let's assume the expectation E is taken over model prediction pθ(x^0∣xt). Denote x^0 as **one-hot** variables and x~0=E[x^0] as the **soft** probability vector.
> 
> * "Direct substitution" can be understood as q(xt−1∣xt,x~0)=q(xt−1∣xt,E[x^0]);
> * The other way you referred to is actually pθ(xt−1∣xt)=Σx^0=1Kpθ(x^0∣xt)q(xt−1∣xt,x^0)=E[q(xt−1∣xt,x^0)].
> 
> These two terms are not equal in general, and the reason is precisely the same as what you have pointed out in your [comment](https://github.com/ehoogeboom/multinomial_diffusion/issues/9#issuecomment-1331795544):
> 
> > Meanwhile, the transition from Eq.(9) to Eq.(10) above is what I feel doubtful. The proportionality may not hold exactly. This is because ...

I see. This really makes sense. Great gratitudes!

**Comment by ehoogeboom** [comment]
Hi both,

Sorry I've been a little absent here, but I agree with all the statements above, well-said. Indeed as was said, we just replaced x_0 with a soft prediction so that p(x_t-1 | x_t) = q(x_t-1 | x_t, E[x_0 | x_t] ) which was a "lucky guess" based on what gaussian diffusion models do. 

Best,
Emiel
